[exporter/file] Add group_by configuration (#31396)
open-telemetry/opentelemetry-collector-contrib: exporter/fileexporter/README.md

+ Support for compressing the telemetry data before exporting.

+ Supports for writing into multiple files, where the file path is determined by a resource attribute.

Please note that there is no guarantee that exact field names will remain stable.

> This intended for primarily for debugging Collector without setting up backends.

Reviewer: Why is this line removed in this PR? Does not seem related to the change.

Author: With every new feature added to this exporter, it becomes less and less of a tool for debugging, so the removal of this line is lightly related to the new functionality added in this pr. That was my reason for removing it (after discussing it in the related issue).

But I'm not going to argue about a line in a readme. If you feel strongly about keeping it (or not removing it in this PR because it's not strictly related to it), I'm happy to add it back.

Reviewer: > after discussing it in the related issue

Thanks, this is a good point. I'm fine with this. :+1:

Otherwise, when using `proto` format or any kind of encoding, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.

## Group by attribute

> By specifying `group_by.resource_attribute` in the config, the exporter will determine a filepath for each telemetry record, by concatenating the value of the resource attribute to the `path` configuration value.

Reviewer 2: Technically not concatenation with the updated design.

```suggestion
By specifying `group_by.resource_attribute` in the config, the exporter will determine a filepath for each telemetry record, by substituting the value of the resource attribute into the `path` configuration value.
```

> The exporter will NOT write files outside of the `path` config value, eg: if `path` is "/data", and the resource attribute value is "../etc/my_config.yaml", then the final path will be sanitized to "/data/etc/my_config.yaml".

Reviewer 2: This example may need to be updated based on the newer design.
